templates/seL4VMParameters.template.h

```diff
@@ -1,27 +1,40 @@
 /*
  * Copyright 2023, DornerWorks
+ * Copyright 2023, Hensoldt Cyber
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
 #pragma once
 
-extern const unsigned long ram_base;
-extern const unsigned long ram_paddr_base;
-extern const unsigned long ram_size;
-extern const unsigned long dtb_addr;
-extern const unsigned long initrd_addr;
-extern const unsigned long entry_addr;
-
-extern const int provide_initrd;
-extern const int generate_dtb;
-extern const int provide_dtb;
-extern const int map_one_to_one;
-extern const int clean_cache;
-
-extern const char *_kernel_name;
-extern const char *_dtb_name;
-extern const char *_initrd_name;
-extern const char *kernel_bootcmdline;
-extern const char *kernel_stdout;
-extern const char *dtb_base_name;
+typedef struct {
+
+    struct {
+        unsigned long phys_base;
+        unsigned long base;
+        unsigned long size;
+    } ram;
+
+    int provide_initrd;
+    int generate_dtb;
+    int provide_dtb;
+    int map_one_to_one;
+    int clean_cache;
+
+    unsigned long dtb_addr;
+    unsigned long initrd_addr;
+    unsigned long entry_addr;
+
+    struct {
+        char const *kernel;
+        char const *initrd;
+        char const *dtb;
+        char const *dtb_base;
+    } files;
+
+    char const *kernel_bootcmdline;
```

Comment:
Is there some specific reason to write it this way and not `const char *` like elsewhere in the code?

Reply:
`const char *` is the sloppy way to write it, `char const *` is more correct actually when it comes to the order.

Reply:
Is that really so? I know the difference between `char const *` and `char * const`, the former being a non-constant pointer to constant character and the latter being a constant pointer to non-constant character. In `const char * const` both the pointer and the character pointed to would be constant. I also know `const char *` and `char const *` are equivalent, but that the latter would be more correct is new to me.

Also if we do simple `grep|wc` for both the seL4 kernel and CAmkES projects, there are roughly 50 occurrences of `const char *` per one occurrence of `char const *`. Should we fix those some day then? And how about other types, should we rewrite `const unsigned long *` as `unsigned long const *`? Does it apply to pointers only, or should we fix `const unsigned long ram_base` to `unsigned long const ram_base`?

Reply:
> Is that really so?

`const` binds to the left. Unless it is the first thing, then it binds to the right. That is why you can define a constant with `const int foo = 42`, which looks a bit nicer than `int const foo = 42`.
For an instance `vm_config_t`, the member don't have to be constants. It can be be used and changed during runtime also to manage a configuration. That why I avoid using a `const` in the beginning. Just for the instance we create from the template, it happens that all data is known at compile time, so whole instance is define as a const to allows it to be put in ro-data.

> ... seL4 kernel and CAmkES projects ... Should we fix those some day then?
In some places it could make sense, e.g. for function parameter types or struct members. When constants are declared, it I'd keep it stating with const. 

>  should we rewrite `const unsigned long *` as `unsigned long const *`
In my opinion, yes. For every pointer it makes sense.

> Does it apply to pointers only, or should we fix `const unsigned long ram_base` to `unsigned long const ram_base`?
For declaring integers declaring a constant `int const foo = 42` looks really odd, so I would not change this.

Reply:
OK, good to learn about your preferences. Perhaps if I toyed with the idea a little bit longer, I might even like it. But that's beside the point. The real reason why I questioned the style choice is, at the risk of repeating myself, shouldn't we follow the existing style of whatever project  we are contributing to? For example, the pattern `if (rvalue == lvalue)` has its merits and those are hard to argue against -- IMO there's only one real reason: for those unfamiliar with the pattern it distracts the attention from the focus point. And then there are sensitive ones, like me, who spot style deviations at a first glance (of course I do accidentally those as well, but that's another story). I can speak only for myself, but I don't do that in order to be able to nag and harass people, but it really throws me off. That's why I suggested perhaps we should stick to existing style, even if it's not the best of the best.

Reply:
@hlyytine, your principles about how we try and follow a consistent style are aligned with mine.

We do have a style guide: https://docs.sel4.systems/processes/style-guide.html and the default assumption is that people will try and follow existing conventions in a file.

> const char * is the sloppy way to write it, char const * is more correct actually when it comes to the order.

How can one approach be more correct over another when they have the same meaning? The [standard](https://www.open-std.org/jtc1/sc22/WG14/www/docs/n1256.pdf) even uses  `const int *ptr_to_constant;` as the first example of a non-constant pointer to a constant value. 

I don't have much spare time to police style on PRs, so I usually make one of two choices:
- Ignore and approve, which reduces the overall level of consistency,
- Engage in the debate and then have less time for reviewing/working on other PRs which then take longer to resolve.

At the end of the day, this is an open source project and the style should serve us towards making it easier to collaborate and maintain, not the other way round. With collaborators from all over the place, it's not really easy to have a tight consistent style without being prepared to rewrite everyone's PRs ourselves.

We have a tool that that both checks style and auto-formats invalid style.  Anything that the tool accepts is acceptable to merge.  The tool roughly tries to implement the style guide I linked above.  Any sufficiently strong opinions for changes in acceptable style can either propose an update to the tool or the style-guide, but style rules must be able to be auto-enforced and applied by the tool for us to consider them.